### ISSUE TITLE
plugin/rewrite: initial request is modified with OPT records

### ISSUE THREAD

**Issue by fturib** [coredns/coredns#2484]
### Context
`CoreDNS v 1.3.1`
`Corefile:`
```
.:53 {
                rewrite edns0 subnet append 24 56
                forward . 10.82.3.15:53
                log
                errors
}
```

### After analysis of the code (credits to @rdrozhdzh)

> I see that rewrite plugin creates OPT record in original request in case if the OPT record was absent from user query. > (https://github.com/coredns/coredns/blob/3533065ed20a7fe14cc00968e3fa8c1beff072b2/plugin/rewrite/edns0.go#L42-L49)

> Thus, when coredns gets response it checks in `func (r *Request) SizeAndDo(m *dns.Msg)` if original query had OPT records. However, since rewrite plugin has modified original query, then SizeAndDo() doesn't know what was in original user query. It supposes that the OPT record with UPD bufsize options came from user.

> Perhaps, rewrite plugin should create a copy of original query before adding OPT record. 

> Also, pay attention, that if coredns responds with OPT record to user query which didn't have OPT RR - this is a violation of RCF https://tools.ietf.org/html/rfc6891

### Initial symptoms (FYI)

CoreDNS returns a packet larger that the regular UDP size in our case
`dig -t TXT <fqdn> +short @<my-resolver-ip> +noedns`

- Client sends query to  CoreDNS  without any EDNS option  41  (i.e.  expect to use  packet size of 512 bytes) 
- CoreDNS inserts  ECS  EDNS option and EDNS option 41 with packet size of 4096 bytes  
- The downstream DNS server responds with  a packet greater than  512 bytes   which eventually forwarded to the client.

**Comment by fturib** [comment]
As it seems simple to modify the initial request, but we need it for replies, and checks ...

Would it be better to keep aside a copy of that initial request (or at least the element that allow to send back properly the response to the client) ?

**Comment by fturib** [comment]
the plugin Rewrite implements a `ResponseReverter` that could be enhanced somehow to ensure to revert also the EXTRA records when needed.
(it is already triggered by a boolean `needRevert`)

**Comment by rdrozhdzh** [comment]
@fturib, I like your idea with `ResponseReverter`. This way we can only store a copy of original `OPT `record without copying entire DNS query message. And `ResponseReverter` will compare the response `OPT` record to original query record.

E.g.  if original `OPT` is `nil` (user didn't send any EDNS data) then `ResponseReverter` will remove entire `OPT` record from response. And if response `OPT` has some options not requested in original request (probably inserted by rewrite plugin) then such options should also be removed from response.

This way we likely could keep `SizeAndDo()` unchanged.

**Comment by fturib** [comment]
@rdrozhdzh : thanks for your comment.
Using ResponseReverter was my first try ... but I guess I got confused by the fact that Copy() of Msg or OPT is not properly working.

After, about the change in SizeAndDo(), I was considering this part of the RFC you pointed:
> 6.2.6.  Support in Middleboxes
>Middleboxes that have additional functionality, such as answering
   queries or acting as intelligent forwarders, SHOULD be able to
   process the OPT record and act based on its contents.  These
   middleboxes MUST consider the incoming request and any outgoing
   requests as separate transactions if the characteristics of the
   messages are different.

But after a second thought, it is maybe REWRITE that is the intelligent part here, and should consider what to do with OPT in both direction.

QUESTION: if the client is sending an empty OPT record (only set the buffer size), if REWRITE is adding an OPT or modifying this OPT to add a code/data -  I mean on the request, for the upstream resolver - Do you expect that the whole modified OPT record return to the client ?

In other words, the modifications on OPT added by REWRITE on the upstream side should be removed from response when sent back to client ?



ASIDE NOTE: I am wondering if we should also verify this part:
> 6.2.1.  Cache Behaviour
> The OPT record MUST NOT be cached.

**Comment by rdrozhdzh** [comment]
> QUESTION: if the client is sending an empty OPT record (only set the buffer size), if REWRITE is adding an OPT or modifying this OPT to add a code/data - I mean on the request, for the upstream resolver - Do you expect that the whole modified OPT record return to the client ?
> 
> In other words, the modifications on OPT added by REWRITE on the upstream side should be removed from response when sent back to client ?

My opinion is that DNS server should not return in response to client any EDNS options which had not been requested by client. Thought, I didn't find a direct statement of this in any RFC. The only close requirement is `6.2.6. Support in Middleboxes` mentioned by you above.

**Comment by fturib** [comment]
I update with a new PR here : #2491.
Changed the behavior, according to comment above.

**Comment by rdrozhdzh** [comment]
From https://tools.ietf.org/html/rfc6891#section-7
```
   Lack of presence of an OPT record in a request MUST be taken as an
   indication that the requestor does not implement any part of this
   specification and that the responder MUST NOT include an OPT record
   in its response.
```
BTW, is there any progress on this issue?
